lib/api/src/lib/stories.ts

```diff
@@ -180,120 +174,162 @@ export const denormalizeStoryParameters = ({
+}
+
+export const transformSetStoriesStoryDataToStoriesHash = (
+  data: SetStoriesStoryData,
   { provider }: { provider: Provider }
-): StoriesHash => {
-  const countByTitle = countBy(Object.values(index.entries), 'title');
-  const input = Object.entries(index.entries).reduce((acc, [id, { title, name, importPath }]) => {
-    const docsOnly = name === 'Page' && countByTitle[title] === 1;
-    acc[id] = {
-      id,
-      kind: title,
-      name,
-      parameters: { fileName: importPath, options: {}, docsOnly },
-    };
-    return acc;
-  }, {} as StoriesRaw);
+) =>
+  transformStoryIndexToStoriesHash(transformSetStoriesStoryDataToPreparedStoryIndex(data), {
+    provider,
+  });
+
+const transformSetStoriesStoryDataToPreparedStoryIndex = (
+  stories: SetStoriesStoryData
+): PreparedStoryIndex => {
+  const entries: PreparedStoryIndex['entries'] = Object.entries(stories).reduce(
+    (acc, [id, story]) => {
+      if (!story) return acc;
+
+      const { docsOnly, fileName, ...parameters } = story.parameters;
+      const base = {
+        title: story.kind,
+        id,
+        name: story.name,
+        importPath: fileName,
+      };
+      if (docsOnly) {
+        acc[id] = {
+          type: 'docs',
+          ...base,
+        };
+      } else {
+        const { argTypes, args, initialArgs } = story;
+        acc[id] = {
+          type: 'story',
+          ...base,
+          parameters,
+          argTypes,
+          args,
+          initialArgs,
+        };
+      }
+      return acc;
+    },
+    {} as PreparedStoryIndex['entries']
+  );
 
-  return transformStoriesRawToStoriesHash(input, { provider, prepared: false });
+  return { v: 4, entries };
 };
 
-export const transformStoriesRawToStoriesHash = (
-  input: StoriesRaw,
-  { provider, prepared = true }: { provider: Provider; prepared?: Story['prepared'] }
+export const transformStoryIndexToStoriesHash = (
+  index: PreparedStoryIndex,
+  {
+    provider,
+  }: {
+    provider: Provider;
+  }
 ): StoriesHash => {
-  const values = Object.values(input).filter(Boolean);
-  const usesOldHierarchySeparator = values.some(({ kind }) => kind.match(/\.|\|/)); // dot or pipe
-
-  const storiesHashOutOfOrder = values.reduce((acc, item) => {
-    const { kind, parameters } = item;
-    const { sidebar = {}, showRoots: deprecatedShowRoots } = provider.getConfig();
-    const { showRoots = deprecatedShowRoots, collapsedRoots = [], renderLabel } = sidebar;
-
-    if (typeof deprecatedShowRoots !== 'undefined') {
-      warnLegacyShowRoots();
-    }
+  const entryValues = Object.values(index.entries);
+  const { sidebar = {}, showRoots: deprecatedShowRoots } = provider.getConfig();
+  const { showRoots = deprecatedShowRoots, collapsedRoots = [], renderLabel } = sidebar;
+  const usesOldHierarchySeparator = entryValues.some(({ title }) => title.match(/\.|\|/)); // dot or pipe
+  if (typeof deprecatedShowRoots !== 'undefined') {
+    warnLegacyShowRoots();
```

Comment:
At what point can we get rid of this? Now?

Reply:
Possibly but didn't want to creep scope